ledger.go

```diff
@@ -1019,426 +938,535 @@ func (l *Ledger) SyncToLatestBlock() { // nolint:gocyclo,gocognit
+					continue
+				}
 
-						if err != nil {
-							continue
-						}
+				sources[src.idx].size = len(chunk)
 
-						sources[src.idx].size = len(chunk)
+				break
+			}
+		}(src)
+	}
 
-						break
-					}
+	chunkWG.Wait() // Wait until all chunks have been received.
 
-					chunkWG.Done()
-				}
+	logger.Debug().
+		Int("num_chunks", len(sources)).
+		Msg(
+			"Downloaded whatever chunks were available to sync to the latest block, and shutted down all " +
+				"workers. Checking validity of chunks...",
+		)
 
-				workerWG.Done()
-			}()
-		}
+	// Check all chunks has been received
+	var diffSize int64
 
-		for _, src := range sources {
-			workers <- src
+	for i, src := range sources {
+		if src.size == 0 {
+			logger.Error().
+				Uint64("target_block_id", latest.Index).
+				Hex("chunk_checksum", sources[i].checksum[:]).
+				Msg("Could not download one of the chunks necessary to sync to the latest block! Retrying...")
+
+			return false
 		}
 
-		chunkWG.Wait() // Wait until all chunks have been received.
-		close(workers)
-		workerWG.Wait() // Wait until all workers have been closed.
+		diffSize += int64(src.size)
+	}
+
+	if _, err := io.CopyN(diffBuffer, chunksBuffer, diffSize); err != nil {
+		logger.Error().
+			Uint64("target_block_id", latest.Index).
+			Err(err).
+			Msg("Failed to write chunks to bounded memory buffer. Restarting sync...")
+
+		return false
+	}
 
-		logger.Debug().
-			Int("num_chunks", len(sources)).
-			Int("num_workers", cap(workers)).
-			Msg(
-				"Downloaded whatever chunks were available to sync to the latest block, and shutted down all " +
-					"workers. Checking validity of chunks...",
-			)
+	logger.Info().
+		Int("num_chunks", len(sources)).
+		Uint64("target_block", latest.Index).
+		Msg("All chunks have been successfully verified and re-assembled into a diff. Applying diff...")
 
-		dispose() // Shutdown all streams as we no longer need them.
+	snapshot := l.accounts.Snapshot()
+	if err := snapshot.ApplyDiff(diffBuffer); err != nil {
+		logger.Error().
+			Uint64("target_block_id", latest.Index).
+			Err(err).
+			Msg("Failed to apply re-assembled diff to our ledger state. Restarting sync...")
 
-		// Check all chunks has been received
-		var diffSize int64
+		return false
+	}
 
-		for i, src := range sources {
-			if src.size == 0 {
-				logger.Error().
-					Uint64("target_block_id", latest.Index).
-					Hex("chunk_checksum", sources[i].checksum[:]).
-					Msg("Could not download one of the chunks necessary to sync to the latest block! Retrying...")
+	if checksum := snapshot.Checksum(); checksum != latest.Merkle {
+		logger.Error().
+			Uint64("target_block_id", latest.Index).
+			Hex("expected_merkle_root", latest.Merkle[:]).
+			Hex("yielded_merkle_root", checksum[:]).
+			Msg("Failed to apply re-assembled diff to our ledger state. Restarting sync...")
 
-				cleanup()
+		return false
+	}
 
-				goto SYNC
-			}
+	// get all the blocks up until latest one
+	var oldestBlockIx uint64 = 1
```

Comment:
Wasn't the objective to perform Snowball sampling and consensus over the last `PruningLimit` number of blocks, rather than just downloading all the latest blocks from `latestVoter`?

Downloading and trusting all the latest blocks from `latestVoter` is quite dangerous.

Reply:
picking block just from the first voter is also dangerous.
i don't think its feasible to do it that way rn. we cannot send all blocks in a single request, so we need to send only block ids and sample those and then download blocks by ids from random peers.
i did it this way since we already relying on one voter by picking latest block by index, so it was logical to get other blocks from same voter.

Reply:
Hm, rather than do it in a single request, I feel we should do it by:

1. If we are out of sync stop consensus.
2. Sample repetitively for the latest blocks contents from peer(s) using Snowball. Let's say we are now in-sync, knowing the block at height H.
3. Sample repetitively for height H+1, and do this for blocks at heights H+PruningLimit.
4. We now have PruningLimit number of blocks. Then continue on with consensus.

Reply:
i don't understand why we should sample for PruningLimit number of blocks after latest one, unless its only to avoid getting already finalized transactions. And IMO its really big time to wait in order for single node to get back in sync.
Won't it be easier to sample for chain of blocks i.e. PruningLimit number of block ids like i suggested earlier?

Reply:
Sampling for a chain of blocks is possible too, though we need to continually be sure to sample the next blocks that proceed to get finalized too in parallel.

If we sample the chain of the last PruningLimit blocks and the majority of nodes only keep the last PruningLimit blocks in their database, we may be unable to fetch the latestBlock.Height - PruningLimit block in time.